### ISSUE TITLE
Limit touching to once per transaction

### ISSUE THREAD

**Issue by dhh** [rails/rails#18606]
If you have a deeply nested hierarchy of objects like so:

```ruby
Comment belongs_to :message, touch: true
Message belongs_to :project, touch: true
Project belongs_to :account, touch: true
```

And you do something like:

```ruby
comments.size # 3
Comment.transaction { comments.each &:touch! }
```

That'll generate 3x touches to each Message, Project, and Account, when in reality they all only need one touch. It would be preferably to have a touch manager of sorts, that ensure that only one touch would happen per transaction.

**Comment by georgeclaghorn** [comment]
I'm interested in looking into this.

**Comment by dhh** [comment]
:metal:

**Comment by sivsushruth** [comment]
Any updates @georgeclaghorn ?

**Comment by georgeclaghorn** [comment]
Still working on it, @sivsushruth.

**Comment by BMorearty** [comment]
Last summer I wrote the [activerecord-delay_touching gem](https://github.com/godaddy/activerecord-delay_touching/) to solve this problem.

Earlier today I began a discussion about it in the [Rails core mailing list](https://groups.google.com/forum/#!topic/rubyonrails-core/5pD0dLhelOE), without knowing that this issue had been opened, to propose that I merge the gem into Rails. My solution was slightly different than your proposal, @dhh, but similar in spirit.

**Comment by dhh** [comment]
Hey @BMorearty, would be great to have you work on this. I believe this should be automatic. You shouldn't have to wrap anything explicitly. Every item in a transaction should only be touched once, since nothing else makes sense – the prior touches were never visible outside of the transaction.

**Comment by BMorearty** [comment]
Sounds good, @dhh. I've got almost all tests passing in a local branch of Rails using my wrapper API. I'll add a few more tests that I thought of, and then convert what I have to an implicit consolidation of touches inside any transaction, rather than explicitly wrapping them.

I hope I'm not stepping on any toes by jumping in on this. I had been planning to submit the PR anyway because the code has been working in production for six months, and only after I started the discussion did I see that you had already created this issue and someone had started some work.

**Comment by georgeclaghorn** [comment]
Not at all, @BMorearty. Thanks for your work! :heart:

**Comment by dhh** [comment]
Awesome! Thanks for your work. 

I'd actually also love to see a generic version of this. Something to ensure that a certain action only happens once per transaction. I have other AR callbacks that behave like touch and can be overly chatty. Maybe there's an abstraction lurking here to deal with those as well?

**Comment by al2o3cr** [comment]
@dhh - there's a similar need for dealing with autosaving associations; issues like #17015 get caused when the saving order gets broken by callback redefinitions. Maybe a mechanism for explicitly saying "X should save after Y" that used `tsort` to decide what order to persist things?

**Comment by BMorearty** [comment]
@dhh I've submitted the PR for this. I modified my code to run automatically in any transaction rather than requiring an explicit call to `delay_touching`. See #18824.

I did not attempt to implement your suggestion for a generic version of this. In the absence of another specific use case to test against, it seemed too abstract for me to try to implement as part of this PR. Besides, this change does quite a bit more than just make sure each touch happens only once in a transaction. It also uses a SQL `WHERE id IN (1, 2, 3)` clause to combine all touches in each table. And to handle cases where `after_touch` callbacks result in additional touches, it keeps iterating until there are no more touches to process.

I kept the code nicely isolated. The benchmark results are pretty good. See the PR for more detail.
